spread/openstack.go

```diff
@@ -0,0 +1,787 @@
+	ListImagesDetail() ([]glance.ImageDetail, error)
+}
+
+type novaComputeClient interface {
+	ListFlavors() ([]nova.Entity, error)
+	ListAvailabilityZones() ([]nova.AvailabilityZone, error)
+	GetServer(serverId string) (*nova.ServerDetail, error)
+	ListServersDetail(filter *nova.Filter) ([]nova.ServerDetail, error)
+	RunServer(opts nova.RunServerOpts) (*nova.Entity, error)
+	DeleteServer(serverId string) error
+}
+
+type openstackProvider struct {
+	project *Project
+	backend *Backend
+	options *Options
+
+	region        string
+	osClient      gooseClient.Client
+	computeClient novaComputeClient
+	networkClient *neutron.Client
+	imageClient   glanceImageClient
+
+	mu sync.Mutex
+
+	keyChecked bool
+	keyErr     error
+}
+
+type openstackServer struct {
+	p *openstackProvider
+	d openstackServerData
+
+	system  *System
+	address string
+}
+
+type openstackServerData struct {
+	Id       string
+	Name     string
+	Flavor   string    `json:"machineType"`
+	Networks []string  `json:"network"`
+	Status   string    `yaml:"-"`
+	Created  time.Time `json:"creationTimestamp"`
+
+	Labels map[string]string `yaml:"-"`
+}
+
+func (s *openstackServer) String() string {
+	if s.system == nil {
+		return s.d.Name
+	}
+	return fmt.Sprintf("%s (%s)", s.system, s.d.Name)
+}
+
+func (s *openstackServer) Label() string {
+	return s.d.Name
+}
+
+func (s *openstackServer) Provider() Provider {
+	return s.p
+}
+
+func (s *openstackServer) Address() string {
+	return s.address
+}
+
+func (s *openstackServer) System() *System {
+	return s.system
+}
+
+func (s *openstackServer) ReuseData() interface{} {
+	return &s.d
+}
+
+const (
+	openstackStaging      = "STAGING"
+	openstackProvisioning = "PROVISIONING"
+	openstackRunning      = "RUNNING"
+	openstackStopping     = "STOPPING"
+	openstackStopped      = "STOPPED"
+	openstackSuspending   = "SUSPENDING"
+	openstackTerminating  = "TERMINATED"
+
+	openstackPending = "PENDING"
+	openstackDone    = "DONE"
+)
+
+func (p *openstackProvider) Backend() *Backend {
+	return p.backend
+}
+
+func (p *openstackProvider) Reuse(ctx context.Context, rsystem *ReuseSystem, system *System) (Server, error) {
+	s := &openstackServer{
+		p:       p,
+		address: rsystem.Address,
+		system:  system,
+	}
+	err := rsystem.UnmarshalData(&s.d)
+	if err != nil {
+		return nil, fmt.Errorf("cannot unmarshal openstack reuse data: %v", err)
+	}
+	return s, nil
+}
+
+func (p *openstackProvider) Allocate(ctx context.Context, system *System) (Server, error) {
+	if err := p.checkKey(); err != nil {
+		return nil, err
+	}
+
+	s, err := p.createMachine(ctx, system)
+	if err != nil {
+		return nil, err
+	}
+
+	printf("Allocated %s.", s)
+	return s, nil
+}
+
+func (s *openstackServer) Discard(ctx context.Context) error {
+	return s.p.removeMachine(ctx, s)
+}
+
+const openstackCloudInitScript = `
+#cloud-config
+runcmd:
+  - echo root:%s | chpasswd
+  - sed -i 's/^\s*#\?\s*\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/' /etc/ssh/sshd_config
```

Comment:
it might be a good idea to comment out existing conflicting configs first, similar to what is suggested here https://github.com/snapcore/spread/pull/155#discussion_r1530108525